sealir-tutorials/ch06_1_typeinfer_ifelse.py

```diff
@@ -0,0 +1,1580 @@
+    push: Callable
+    tos: Callable
+
+@dataclass(frozen=True)
+class LLVMValue: ...
+
+
+@dataclass(frozen=True)
+class IOState(LLVMValue): ...
+
+
+@dataclass(frozen=True)
+class SSAValue(LLVMValue):
+    value: int
+
+@dataclass(frozen=True)
+class PackedValues(LLVMValue):
+    values: tuple[LLVMValue, ...]
+
+    @classmethod
+    def make(cls, *args) -> PackedValues:
+        return cls(args)
+
+    def __post_init__(self):
+        for v in self.values:
+            # TODO: Assert properly here
+            pass
+            # assert isinstance(v, (SSAValue, IOState)), type(v)
+
+    def __getitem__(self, idx) -> SSAValue:
+        return self.values[idx]
+
+    def __len__(self) -> int:
+        return len(self.values)
+
+def determine_arity(root: ase.SExpr) -> int:
+    node = root
+    match node:
+        case rg.Func(args=rg.Args() as args):
+            return len(args.arguments)
+        case _:
+            raise TypeError(node._head)
+
+class NbOp_Not_Int64(NbOp_Base):
+    operand: SExpr
+
+class Backend:
+    def lower_type(self, ty: NbOp_Type):
+        match ty:
+            case NbOp_Type("Int64"):
+                return ir.IntType(64)
+        raise NotImplementedError(f"unknown type: {ty}")
+
+    def get_mlir_type(self, seal_ty):
+        match seal_ty.name:
+            case "Int64":
+                return self.i64
+            case "Float64":
+                return self.f64
+
+    def lower(self, root: rg.Func, argtypes):
+        self.context = context = ir.Context()
+        self.loc = loc = ir.Location.unknown(context=context)
+        self.module = module = ir.Module.create(loc=loc)
+
+        # f32 = ir.F32Type.get(context=context)
+        self.f64 = f64 = ir.F64Type.get(context=context)
+        self.i32 = i32 = ir.IntegerType.get_signless(32, context=context)
+        self.i64 = i64 = ir.IntegerType.get_signless(64, context=context)
+        self.boo = boo = ir.IntegerType.get_signless(1, context=context)
+
+        module_body = ir.InsertionPoint(module.body)
+        input_types = tuple([self.get_mlir_type(x) for x in argtypes])
+        # TODO: Fix this
+        output_types = (self.f64,)
+
+        with context, loc, module_body:
+            fun = func.FuncOp("func", (input_types, output_types))
+            fun.attributes["llvm.emit_c_interface"] = ir.UnitAttr.get()
+            const_block = fun.add_entry_block()
+            fun.body.blocks.append(*[], arg_locs=None)
+            func_block = fun.body.blocks[1]
+
+        constant_entry = ir.InsertionPoint(const_block)
+        function_entry = ir.InsertionPoint(func_block)
+
+        region_args = []
+
+        @contextmanager
+        def push(arg_values):
+            region_args.append(tuple(arg_values))
+            try:
+                yield
+            finally:
+                region_args.pop()
+
+        def get_region_args():
+            return region_args[-1]
+        
+        def lower_expr(expr: SExpr, state: ase.TraverseState):
```

Comment:
This closure is refactored into a proper method in later commit. You'll need to rebase on the latest commit for ch4.